### ISSUE TITLE
[bench request]: Add blink-4b (Qwen3.5-4B + LoRA, one-pass option probabilities, TypeSafe wire format, self-hosted)

### ISSUE THREAD

**Issue by thegovind** [fstandhartinger/jevbench#81]
## The system

Please evaluate [blink-4b](https://huggingface.co/thegovind/blink-4b) as a self-hosted entrant at
revision `c243abdd2d79478db510dcdc60a88087c18547c0` (tag `v1.0`). It's merged weights: a uniform average of three LoRA fine-tunes of
`Qwen/Qwen3.5-4B` (Apache-2.0). Each batch of questions takes one forward pass and returns native
probabilities over the offered options for `choice`, `noul` and `score`; no text is generated.
The weights are for non-commercial research and evaluation (`LICENSE.md` in the model repo).
`serve.py`, `blink.py` and the Dockerfile are Apache-2.0.

## How to run it

Download the pinned revision and run the author's server locally:

```sh
pip install "torch==2.13.0" "transformers==5.17.0" "flash-linear-attention==0.5.2" "accelerate>=1.1.0" safetensors huggingface_hub
hf download thegovind/blink-4b --revision c243abdd2d79478db510dcdc60a88087c18547c0 --local-dir blink-4b
python blink-4b/serve.py --model ./blink-4b --port 8000
```

In another terminal: `curl -s http://127.0.0.1:8000/healthz`. Or, from the downloaded folder:

```sh
cd blink-4b
docker build -t blink-4b . && docker run --rm --gpus all -p 127.0.0.1:8000:8000 blink-4b
```

The stock `typesafe` adapter can use `http://127.0.0.1:8000`; `TYPESAFE_API_KEY` may be any
value. No adapter changes or code PR are needed. Please run one request at a time on your own
machine. We don't operate an endpoint or receive held-out or sealed items.

`/healthz` reports `weights_verified` (files listed in `weights.sha256` are hashed; a mismatch
stops startup), `warmup.repeat_identical` (two identical warm-up answers), `kernels` (fast
linear-attention or slower fallback), `versions` and `hub_offline`. Serving the downloaded
folder or Docker image enables Hugging Face offline mode before loading (`hub_offline: true`);
the server doesn't otherwise restrict network access. It needs one CUDA GPU and
flash-linear-attention for the fast path; bf16 weights are 8.4 GB.

Temperature stays at 1.0 (not fitted). The server keeps the client's option order and uses
no test-time ensembling. For `choice`, `confidence = (p_max − 1/K)/(1 − 1/K)`. Limits: 255
options per choice, 2–10 score levels, 131,072 input tokens per question and 512 questions
per request. Over-limit requests return HTTP 422 with a reason; nothing is truncated.

A fresh install with these pins served blink-4b; through the stock adapter, all 231
public-item responses matched our development run byte for byte: easy 48/48, standard 71/72,
hard 80/111. That's a setup check, not an official score. The Dockerfile uses the same torch,
transformers and flash-linear-attention pins;
we built and ran it on CPU with a small stand-in model, checking health, answers and HTTP 422
limits. We did not run that image with blink-4b on a GPU. JevBench's client measured serial
loopback timing on our machine: easy/standard p50 64 ms; hard p50 67 ms, p95 206 ms. For cost,
use Qwen3.5-4B's public list price of $0.03 per million input tokens; the server reports
`usage.input_tokens` and 0 output tokens on every response.

## Disclosures

- Training labels came from code-generated decision and exact-probability worlds, code-output
  and causal reasoning. Judge-style rows cover GSM8K-train solution checks, Dolly-15k routing
  and program-answer checks. blink-4b has no chess rows. Public dataset train/dev splits are
  listed on the [model card](https://huggingface.co/thegovind/blink-4b).
  One teacher, `Qwen/Qwen3.8-27B` in thinking mode, wrote documents and typed questions. We kept
  them only when a blind re-solve by that same teacher agreed, not as independent verification.
  We never called Jev's hosted API or used any of its outputs.
- JevBench's published tier descriptions and our public-item results informed the authored
  families: long policy, temporal/numeric, multi-hop, ambiguity/abstain, trade-off, traps,
  judge-style answer checking and routing. An audit of the final training mixture checked
  every row, including teacher-written rows and base-model anchors, against all 231 public
  items using normalised strings of at least 30 characters and 13-word passages. No training
  row shared text with a public item.
- All 231 public items were scored on each candidate checkpoint; JevK5's 65 hand-written hard
  items (Apache-2.0) were a second selection set. We chose the released average with those
  development results. JevK5 was a reference model in our harness and part of one weight-soup
  experiment we dropped; **none of its weights are in blink-4b**.

Please judge held-out and sealed items in your own run. We claim no official score, rank or
parity from these public-item results.

**Comment by fstandhartinger** [comment]
Thanks — this is logged for reachability checks and the JevBench measurement queue. We’ll only make benchmark claims after our own run.

- Harold

**Comment by thegovind** [comment]
I edited the issue description to spell out the final-mixture overlap audit's exact-text and question-passage checks, prompt-template exclusion, and limits. The reproduction check matched decisions, probabilities and token usage on the public items; only the reported model name differed.

**Comment by fstandhartinger** [comment]
One question before we finish: which revision did you mean?

The request pins `c243abdd2d79478db510dcdc60a88087c18547c0` and calls it tag `v1.0`. On the Hub,
`refs/tags/v1.0` points at `f229a9fc…` and `main` is `443889f4…`; the two commits the API lists for the
branch are `656d44da` ("blink v1.0", 16:35 UTC) and `443889f4` ("Card and licence corrections", 18:04 UTC).
`c243abdd…` does resolve, and `serve.py`, `blink.py` and `config.json` are identical at all of them, so
nothing about the model changes — but we would rather the row names the revision you intended.

We ran the revision your request names, with your own `serve.py` on the stack your Dockerfile pins, one
request at a time, offline in a network-disabled container. `weights_verified` was true and
`warmup.repeat_identical` held.

- Harold

**Comment by thegovind** [comment]
Thanks for running the pin offline, one request at a time, and checking `weights_verified` and `warmup.repeat_identical`. We meant **`c243abdd2d79478db510dcdc60a88087c18547c0`**; please keep the row on it.

`v1.0` is an annotated tag: `f229a9fc…` is the tag object, while `refs/tags/v1.0^{}` and the Hub’s `revision=v1.0` both resolve to `c243abdd…`. Our publish script squashed main into `656d44da` but tagged the head it had read before the squash landed. Those commits have the same 20 files byte-for-byte. `443889f4` changes only the card and licence text; the weights and serving files match across all three. We’ll leave the tag in place so the pin stays stable.
